server/src/com/cloud/network/element/VirtualRouterElement.java

```diff
@@ -775,17 +775,24 @@ public boolean savePassword(final Network network, final NicProfile nic, final V
         }
 
         final VirtualMachineProfile uservm = vm;
-
         final DataCenterVO dcVO = _dcDao.findById(network.getDataCenterId());
         final NetworkTopology networkTopology = networkTopologyContext.retrieveNetworkTopology(dcVO);
 
-        // If any router is running then send save password command otherwise
-        // save the password in DB
+        if (!network.isRedundant()) {
+            return networkTopology.savePasswordToRouter(network, nic, uservm, routers.get(0));
+        }
+
+        Boolean applied;
         for (final VirtualRouter router : routers) {
-            if (router.getState() == State.Running) {
-                return networkTopology.savePasswordToRouter(network, nic, uservm, router);
-            }
+                if (router.getState() == State.Running && router.getRedundantState() == VirtualRouter.RedundantState.MASTER) {
+                    applied = networkTopology.savePasswordToRouter(network, nic, uservm, router);
+                    if (!applied) {
+                        return false;
+                    }
+                    return true;
+                }
         }
+        //save password
```

Comment:
save password should be moved to a seperate method, as password should be saved to user_vm_details table if it is not applied to VR, at around line 773.